segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java

```diff
@@ -105,10 +111,13 @@
      * @param attributeIndexFactory    The AttributeIndexFactory to use to create Attribute Indices.
      * @param writerFactory            The WriterFactory to use to create Writers.
      * @param storageFactory           The StorageFactory to use to create Storage Adapters.
+     * @param createExtensions            A Function that, given an instance of this class, will create the set of
+     *                                 {@link SegmentContainerExtension}s to be associated with that instance.
      * @param executor                 An Executor that can be used to run async tasks.
      */
     StreamSegmentContainer(int streamSegmentContainerId, ContainerConfig config, OperationLogFactory durableLogFactory, ReadIndexFactory readIndexFactory,
-                           AttributeIndexFactory attributeIndexFactory, WriterFactory writerFactory, StorageFactory storageFactory, ScheduledExecutorService executor) {
+                           AttributeIndexFactory attributeIndexFactory, WriterFactory writerFactory, StorageFactory storageFactory,
+                           SegmentContainerFactory.CreateExtensions createExtensions, ScheduledExecutorService executor) {
```

Comment:
One clarification that we might want to make in the documentation (e.g., javadocs) is that these extensions are loaded at construction time (I think that's what you mean with "follows the same lifecycle of the container") and not dynamically. I can't load an extension while the container is running, if I get it right.

One thing I'm wondering about is whether it is possible that an extension is loaded in an segment store instance and not in another. Say that the container changes ownership and the new owner is not running the extension. Is this a valid scenario even?

Reply:
The "loading" is hard coded in `ServiceBuilder` and is not driven by configuration. So as long as all Segment Store instances run the same code, we are fine. If they don't, we have much bigger problems to worry about.

Reply:
How are upgrades supposed to work? Say I introduce an extension in a given version _X_ and I want to upgrade my Pravega cluster from version _X - 1_ to _X_. My cluster has data already and I don't want to wipe it out. Have you thought about it or do we need to figure this out?

Reply:
This has nothing to do with data as it doesn't manipulate anything that the container stores or accesses. Extensions only use external container APIs and they have no access to internals.

Take for example the Tables extension: it doesn't touch the internals of the container but it add another layer on top of that that allows us to expose data in specific segments in a specified format (dictated by the extension); nothing is changed underneath that.

Reply:
I see your point, but I'm still wondering about rolling upgrades when adding an extension based on this comment:

> as long as all Segment Store instances run the same code, we are fine

which indicates that we can't do a rolling upgrade of the segment store when upgrading to a version that has a new extension. Maybe that's ok because a feature can't really be exercised when only some instances have it. Take tables as an example. An application can't really use when some segment store instances are missing the new code. Either we have a way to spot when all instances are running it to enable the extension or we require a planned outage. Also, extensions are expected to be added infrequently, so we can plan on which versions to introduce them.